### PR TITLE
IMP: add source-tracking to Usage API

### PR DESCRIPTION
**Pull request by thermokarst**

### CODE REVIEW

**Comment by thermokarst** [comment]
The _rationale_ for this PR is discussed here: https://github.com/qiime2/q2cli/pull/239#issuecomment-627462020

File: qiime2/sdk/tests/test_usage.py [hunk 1]
```diff
@@ -359,4 +359,5 @@ def test_merge_metadata_one_input(self):
 class TestScopeRecord(TestCaseUsage):
     def test_invalid_assert_has_line_matching(self):
         with self.assertRaisesRegex(TypeError, 'should be a `callable`'):
-            usage.ScopeRecord('foo', assert_has_line_matching='spleen')
+            usage.ScopeRecord('foo', 'value', 'source',
+                              assert_has_line_matching='spleen')
```
Comment (andrewsanchez) [inline qiime2/sdk/tests/test_usage.py:363]: What is 'spleen'?

Comment (thermokarst) [reply qiime2/sdk/tests/test_usage.py:363]: That's a pretty deep question!

In this case - nothing, just a placeholder value to appease the constructor.

**Comment by thermokarst** [comment]
Hey @andrewsanchez - I worked in a few stubs and marked some misc TODOs from our convo with @ebolyen last week. Would you mind addressing the remaining TODOs, etc? Thanks!

File: qiime2/sdk/tests/test_usage.py [hunk 2]
```diff
@@ -34,12 +34,16 @@ def test_basic(self):
...
-        self.assertEqual('comment', obs4['type'])
-        self.assertEqual('action', obs5['type'])
+        self.assertEqual('init_data', obs1['source'],
+                         records[obs1['ref']].source)
+        self.assertEqual('init_data', obs2['source'],
+                         records[obs2['ref']].source)
+        self.assertEqual('init_data', obs3['source'],
+                         records[obs3['ref']].source)
+        self.assertEqual('comment', obs4['source'])
+        self.assertEqual('action', obs5['source'])
```
Comment (andrewsanchez) [inline qiime2/sdk/tests/test_usage.py:46]: > # TODO: refactor tests that use this to compare the `type` to the record.source

Is this what you had in mind?

**Comment by andrewsanchez** [comment]
> Hey @andrewsanchez - I worked in a few stubs and marked some misc TODOs from our convo with @ebolyen last week. Would you mind addressing the remaining TODOs, etc? Thanks!

@thermokarst I pushed up my WIP with a couple of comments inline.

File: qiime2/sdk/tests/test_usage.py (hunk 2 above)
Comment (thermokarst) [reply qiime2/sdk/tests/test_usage.py:46]: ❤️  yep!

File: qiime2/core/testing/examples.py [hunk 3]
```diff
@@ -193,6 +194,7 @@ def identity_with_metadata_column_get_mdc(use):
 
 def identity_with_metadata_column_from_factory(use):
     ints = use.init_data('ints', ints1_factory)
+    # TODO: refactor to use `init_metadata` && `get_metadata_column`
```
[the remark this answers is not in the document]
Comment (thermokarst) [reply qiime2/core/testing/examples.py]: Oh duh, yeah - the whole example is pointless now. Remove away! Thanks!

File: qiime2/sdk/tests/test_usage.py [hunk 4]
```diff
@@ -359,4 +363,29 @@ def test_merge_metadata_one_input(self):
...
+            use.init_data('name', lambda: object)
+
+        with self.assertRaisesRegex(ValueError, 'not all .* Artifacts'):
+            use.init_data('name', lambda: [object])
+
+        with self.assertRaisesRegex(TypeError, 'must be of type Metadata'):
+            use.init_metadata('name', lambda: object)
+
+        with self.assertRaisesRegex(ValueError, '`ref` must match'):
+            use.get_metadata_column('name', Metadata, 'a')
```
Comment (andrewsanchez) [inline qiime2/sdk/tests/test_usage.py:384]: Is there a reason not to change the signature of `Usage.get_metadata_column` to something like `get_metadata_column(self, column_name, record)` and pass that through as the `ref`?

Comment (thermokarst) [reply qiime2/sdk/tests/test_usage.py:384]: Great idea - let's do that.

File: qiime2/sdk/usage.py [hunk 5]
```diff
@@ -270,17 +285,21 @@ def _assert_has_line_matching_(self, ref, label, path, expression):
         raise NotImplementedError
 
     def get_result(self, ref):
-        return self._get_record(ref)
+        record = self._get_record(ref)
+        source = record.source
+        if source != "action":
+            raise TypeError(f"source == {source} but must be 'action'")
```
Comment (thermokarst) [inline qiime2/sdk/usage.py:291]: Please use single quotes (sorry for the nitpick)

Comment (thermokarst) [inline qiime2/sdk/usage.py:291]: Also, please use the `%-formatting` style in this module, for consistency.

File: qiime2/sdk/usage.py [hunk 6]
```diff
@@ -346,13 +372,37 @@ def _assert_has_line_matching_(self, ref, label, path, expression):
...
+                                 (factory, result_type))
+
+        return result
+
+    def _init_metadata_(self, ref, factory):
+        data = factory()
+        if not isinstance(data, Metadata):
+            raise TypeError(
+                f"data must be of type Metadata: type(data) == {type(data)}")
+        return data
```
Comment (thermokarst) [inline qiime2/sdk/usage.py:395]: Can you unify the approach used here with the approach above in `_init_data`? They should preferably use the same variable and formatting conventions as above.

Also, please use %-style string formatting, for consistency. Thanks!

File: qiime2/sdk/usage.py [hunk 7]
```diff
@@ -346,13 +372,37 @@ def _assert_has_line_matching_(self, ref, label, path, expression):
...
+            if not all([isinstance(i, sdk.Artifact) for i in result]):
+                raise ValueError('Factory (%r) returned a %s where not all '
+                                 'elements were Artifacts.' %
+                                 (factory, result_type))
+
+        return result
+
+    def _init_metadata_(self, ref, factory):
+        data = factory()
+        if not isinstance(data, Metadata):
```
Comment (thermokarst) [inline qiime2/sdk/usage.py:392]:
```suggestion
        if not isinstance(data, metadata.Metadata):
```

File: qiime2/sdk/usage.py [hunk 8]
```diff
@@ -11,6 +11,7 @@
 import types
 
 from qiime2 import sdk
+from ..metadata import Metadata
```
Comment (thermokarst) [inline qiime2/sdk/usage.py:14]:
```suggestion
from qiime2 import sdk, metadata
```

**Pull request merged**